docarray/doc_index/backends/hnswlib_doc_index.py

```diff
@@ -0,0 +1,400 @@
+import hashlib
+import os
+import sqlite3
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Generic,
+    List,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+import hnswlib
+import numpy as np
+
+import docarray.typing
+from docarray import BaseDocument, DocumentArray
+from docarray.doc_index.abstract_doc_index import (
+    BaseDocumentIndex,
+    FindResultBatched,
+    _Column,
+)
+from docarray.proto import DocumentProto
+from docarray.utils.filter import filter as da_filter
+from docarray.utils.find import FindResult
+from docarray.utils.misc import is_np_int, torch_imported
+
+if TYPE_CHECKING:
+
+    def composable(fn):  # type: ignore
+        return fn
+
+else:
+    # static type checkers do not like callable objects as decorators
+    from docarray.doc_index.abstract_doc_index import composable
+
+TSchema = TypeVar('TSchema', bound=BaseDocument)
+T = TypeVar('T', bound='HnswDocumentIndex')
+
+HNSWLIB_PY_VEC_TYPES = [list, tuple, np.ndarray]
+if torch_imported:
+    import torch
+
+    HNSWLIB_PY_VEC_TYPES.append(torch.Tensor)
+
+
+class HnswDocumentIndex(BaseDocumentIndex, Generic[TSchema]):
+    def __init__(self, db_config=None, **kwargs):
```

Comment:
why `**kwargs` here ? could we do smth so that the docstring are updatrd accordingly ? (could do in another PR but would need a ticket)

Reply:
It is convenience for the user.
Instead of creating a `config` object they can pass all the config params individually, as **kwargs, and it will just work.

Otherwise the user would have to import the config class (differently for each backend), instantiate it with some settings, and pass it, which is cumbersome.

But it can also be useful in certain settings, e.g. if a user want to clearly define all config setting globally. So both options exist.